### ISSUE TITLE
About DXZ.py

### ISSUE THREAD

**Issue by solub** [seanlaw/avm#1]
Hi, 

First off, let me thank you for this implementation of DXZ, it is very rare to find a Python version of this algorithm and yours is instructive.

I have tried to output the exact cover of a relatively simple polyomino tiling problem with it but for some reason the computation is very long (+40 minutes) and it seems to make my browser (Jupyter Notebook) crash before it could converge towards a solution.

It is strange because it takes around 140 seconds to find all the 99392 solutions to that same problem using Donald Kunth's DLX algorithm on my computer (simple Python implementation). 

I am pretty sure I am setting up the input matrix correctly but posting the procedure here as a precaution:

**Tiling Problem:** _Find all permutations of 10 given polyominos within a 5x8 grid_

```
#hard-coding the shapes in 1D (list of cell indices starting from the top-left corner on a 5*8 grid)
polyominos = [[0, 5, 6, 10], [0, 5, 6, 10], 
              [0, 5, 10, 11], [0, 5, 10, 11],
              [0, 1, 6, 7], [0, 1, 6, 7],
              [0, 1, 2, 3], [0, 1, 2, 3],
              [0, 1, 5, 6], [0, 1, 5, 6]]

N = len(polyominos) #number of polyominos -> 10 (5 pairs)
W, H = 5, 8 #grid dimensions
```

Prepare the input matrix with a `Matrix` class (_works well with DLX_), that:

- takes a list of polyominos
- compute their rotations + reflections
- shift their position cell by cell (0, 1, 2, 3, ...) within the grid
- convert these displacements to binary arrays
- stack these arrays in a matrix (2D numpy array)

```
class Matrix(object):
    def __init__(self, width, height, polyominos):
        self.W = width
        self.H = height
        self.pminos = polyominos
        self.rows = []
        self.initialize()
        
        
    def getArray(self):
        return np.array(self.rows).astype('u1')
        
        
    def initialize(self):
        
        for idp, p in enumerate(self.pminos):
            mat = self.toBinaryMatrix(p)
            transformations = self.transform(mat)
            for t in transformations:
                maxn = self.W*self.H - max(t) #maximum displacement index
                maxw = self.W - max(map(lambda f: f%self.W, t)) #maximum width of transformation
                for n in range(maxn):
                    if n%self.W >= maxw:
                        continue

                    shift = [i+n for i in t]
                    
                    a = np.zeros(len(self.pminos))
                    np.put(a, idp, 1)
                    
                    b = np.zeros(self.W*self.H)
                    np.put(b, shift, 1)
                    
                    row = np.concatenate((a, b))
                    self.rows.append(row)
                
        
    def toBinaryMatrix(self, ids):    

        w = max(map(lambda f: f%self.W, ids))
        h = max(map(lambda f: f//self.W, ids))

        return [[1 if x + y * self.W in ids else 0 for x in range(w+1)] for y in range(h+1)]


    def transform(self, matrix):

        arrays = set()

        for i in range(4):
            rot90 = list(zip(*matrix[::-1]))
            vflip = rot90[::-1]
            hflip = [a[::-1] for a in rot90]

            for t in (rot90, vflip, hflip):
                arrays.add(self.flatten(t))

            matrix = rot90

        return arrays


    def flatten(self, matrix):

        arr = [ix+(iy*self.W) for iy, row in enumerate(matrix) for ix, x in enumerate(row) if x == 1]

        return frozenset(arr)
```

Input the matrix into the DXZ solver and run it:

```
mat = Matrix(W, H, polyominos).getArray() 
csc = csc_matrix(mat)
pieces = {0: 'A', 1: 'B', 2: 'C', 3: 'D', 4: 'E', 5: 'F', 6: 'G', 7: 'H', 8: 'I', 9: 'J'} #10 polyominos from column 0 to 9
dxz = DXZ(csc, primary_idx=pieces.keys())
dxz.search() # -> takes for ever !
```

Do you think I am missing something or could it be that your implementation has a flaw ?

Respectfully,

**Comment by seanlaw** [comment]
@solub Unfortunately, it's been way too long since I've touched this code so I'm not certain that I'll be able to help much. Did you also take a look at the `dlx.py` implementation? If I remember correctly, `dlx.py` is based on the Knuth algorithm while `dxz.py` was based on [this later paper](https://aaai.org/ocs/index.php/AAAI/AAAI17/paper/download/14907/13855) that tries to leverage decision diagrams for storage and caching. Frankly, I can't recall if I actually got DXZ working. Luckily, both `dlx.py` and `dxz.py` come with a `__main__` section at the bottom that provides examples of how things work. 

This was meant to be a fun side project for trying to enumerate positions for [this semi-related game](https://www.smartgames.eu/uk/one-player-games/anti-virus-mutation) ("A.V.M. stands for "Anti-Virus Mutation"). I was never able to compute all of the placements even after 2 weeks so I stopped working on it.

**Comment by solub** [comment]
Thank you for the swift reply.

I didn't have a look at the `dlx.py` file but after some comparison testing with another DLX implementation it seems the dxz.py takes much longer than it should. In the original paper, the authors mention a speed-up of several orders of magnitude while in this case, `dxz.py` turns out to be slower than regular DLX. Unfortunately I don't have the ability to check what could possibly be the reason behind this. 

If had the time, I would probably try to create a custom ZDD class and combine it with your DLX implementation instead of relying on Graphillion.

**Comment by seanlaw** [comment]
Graphillion is a reasonably good package. I could never understand the original ZDD papers so I would be interested in any good examples/resources that you came across.

**Comment by solub** [comment]
Sure, I will let you know !

So far I could find [this](https://github.com/travis14/math108-zdd-final-project/blob/master/ZDD.py) Python implementation of a ZDD and also a C++ [version](https://github.com/ekmett/guanxi/blob/master/cbits/dxz.cpp) of DXZ. I also stumbled on an OCaml library called [Combine](https://www.lri.fr/~filliatr/combine/) but I haven't figured out yet if the ZDD module is used in combination with DLX or if is a separate interface.

**Comment by seanlaw** [comment]
IIRC, the person who made Graphillion also works/worked for the same person who created/popularized ZDDs

**Comment by solub** [comment]
Don't get me wrong, I do believe Graphillion is a great library. I am just wondering if it is the right tool for a custom implementation of DXZ.

**Comment by seanlaw** [comment]
Agreed. My lack of understanding of ZDD led me to the only reliably pip installable implementation. I can certainly see that using Graphillion is overkill for what is needed for DXZ.

**Comment by seanlaw** [comment]
I just wished that I had other uses for Graphillion where I could use it to its full potential.
